### PR TITLE
Creating a Simple Real-Time Chat App using React and Socket.io with E2E Encryption

### PR DESCRIPTION
**Pull request by ephnjor2021**
…2E Encryption

Article uploaded

### Article checklist
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

### CODE REVIEW

**Comment by ephnjor2021** [comment]
This solves #1749

**Comment by zolomohan** [comment]
thanks, @WanjaMIKE. will review it tonight.

File: articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md [hunk 1]
```diff
@@ -0,0 +1,832 @@
...
+**Step 2**
+
+Next will be to install the dependencies our server will require by running the following commands:
+
+```bash
+$ npm i socket.io express cors colors
+$ npm i -D nodemon
+```
+
+Then, we code the file `dummyuser.js` as below:
```
Comment (zolomohan) [inline articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md:56]: Before you ask the reader to create a file and add the code below, give a brief one-line intro on what the file will contain and will be used for.
Then, you can explain it in detail.

File: articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md [hunk 2]
```diff
@@ -0,0 +1,832 @@
...
+
+The `join_User()` function adds the user to the array of users already declared in the above code. It consists of three keys *id, a username*, and a *room name*, where the room name tells the user the room or group belongs.
+
+The `get_Current_User()` function will take the id of the particular user and return its user object.
+
+In the `user_Disconnect()` function, if a user disconnects or leaves the chat, the function accepts a *user id* and deletes the user object from the array users.
+
+**Step 3**
+
+In this stage, we'll code the file `server.js` as below:
```
Comment (zolomohan) [inline articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md:104]: Brief intro about the file.

File: articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md [hunk 3]
```diff
@@ -0,0 +1,832 @@
...
+
+```bash
+$ npx create-react-app chatfrontend
+$ cd chatfrontend
+$ npm i node-sass react-redux react-router-dom redux socket.io-client aes256
+```
+
+**Step 2**
+
+Next, we will create a file `/src/index.js` as below:
```
Comment (zolomohan) [inline articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md:212]: Brief intro about the file.

File: articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md [hunk 4]
```diff
@@ -0,0 +1,832 @@
...
+  <Provider store={store}>
+    <App />
+  </Provider>,
+  document.getElementById("root")
+);
+```
+
+In the above code, we have added `redux` and imported `reducers` from the file `/store/reducer/index.js`
+
+Next we create a file `/store/action/index.js` as below:
```
Comment (zolomohan) [inline articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md:234]: Brief intro about the file.

File: articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md [hunk 5]
```diff
@@ -0,0 +1,832 @@
...
+    payload: {
+      encrypt,
+      text,
+      cypher,
+    },
+  };
+};
+```
+
+Then create a file `/store/reducer/index.js` as below:
```
Comment (zolomohan) [inline articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md:249]: Brief intro about the file.

File: articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md [hunk 6]
```diff
@@ -0,0 +1,832 @@
...
+      return state;
+  }
+};
+```
+
+In the above code snippets, we add [redux](https://www.npmjs.com/package/redux) into React App and then create an action by the name `process`. The action will help send and receive incoming and outgoing messages respectively to the file `aes.js`, which will encrypt and decrypt messages.
+
+**Step 3**
+
+Next, let's create the file `App.js` as below:
```
Comment (zolomohan) [inline articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md:279]: Brief intro about the file.

File: articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md [hunk 7]
```diff
@@ -0,0 +1,832 @@
...
+        text: `${p_user.username} has left the chat`,
+      });
+    }
+  });
+});
+```
+
+In the above `server.js` code, we start by importing the modules and functions from the file `dummyuser.js`. The code listens on port *8000* and initializes the socket.
+
+After initializing the socket, the two functions listed below will be called.
```
Comment (zolomohan) [inline articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md:184]: Technically, the function you pass to these event listeners will be called. Please mention the event listeners.

File: articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md [hunk 8]
```diff
@@ -0,0 +1,832 @@
...
+        </Switch>
+      </div>
+    </Router>
+  );
+}
+
+export default App;
+```
+
+In the above, we add routes and importing the components. We render home components and get `username` and `roomname` from the routes on the base URL.
```
Comment (zolomohan) [inline articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md:325]: it's not clear on what components you are importing.

File: articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md [hunk 9]
```diff
@@ -0,0 +1,832 @@
...
+
+$backgroundColor: #282b34;
+$redColor: #ff1e56;
+$yellowColor: #ffac41;
+$greyColor: #2d343e;
+```
+
+**Step 4**
+
+Next will be coding the file `/home/home.js` as below:
```
Comment (zolomohan) [inline articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md:372]: Brief intro about the file.

File: articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md [hunk 10]
```diff
@@ -0,0 +1,832 @@
...
+    &:hover {
+      cursor: pointer;
+    }
+  }
+}
+```
+
+**Step 5**
+
+Next will be to code the file `/chat/chat.js` as shown below:
```
Comment (zolomohan) [inline articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md:458]: Brief intro about the file.

File: articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md [hunk 11]
```diff
@@ -0,0 +1,832 @@
...
+        ></input>
+        <button onClick={sendData}>Send</button>
+      </div>
+    </div>
+  );
+}
+export default Chat;
+```
+
+In the above code, we take the user's input and pass it to the action `process`, and then the data will be passed to `aes` function for encryption. Then the encrypted data is sent to `socket.on("chat")`. Also, if the message is received, it'llill be passed to `aes` function for decryption.
```
Comment (zolomohan) [inline articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md:557]:
```suggestion
In the above code, we take the user's input and pass it to the action `process`, and then the data will be passed to the `aes` function for encryption. Then the encrypted data is sent to `socket.on("chat")`. Also, if the message is received, I'll be passed to the `aes` function for decryption.
```

File: articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md [hunk 12]
```diff
@@ -0,0 +1,832 @@
...
+        cursor: pointer;
+      }
+    }
+  }
+}
+```
+
+**Step 6**
+
+Next will be to create the file `aes.js` as below:
```
Comment (zolomohan) [inline articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md:689]: Brief intro about the file.

File: articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md [hunk 13]
```diff
@@ -0,0 +1,832 @@
...
+  if (cipher.startsWith(username)) {
+    return cipher;
+  }
+
+  var decrypted = aes256.decrypt(secret_key, cipher);
+  return decrypted;
+};
+```
+
+In the above code, we import `aes256` from `aes` module and write a function where the incoming encrypted message is decrypted and encrypts the outgoing message. Note the welcome user message is not to be encrypted.
```
Comment (zolomohan) [inline articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md:714]:
```suggestion
In the above code, we import `aes256` from the `aes` module and write a function where the incoming encrypted message is decrypted and encrypts the outgoing message. Note the welcome user message is not to be encrypted.
```

Comment (zolomohan) [reply articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md:714]: You're writing two functions, not just one.

File: articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md [hunk 14]
```diff
@@ -0,0 +1,832 @@
...
+
+**Step 7**
+
+Next will be creating the `/process/process.js` as below:
```
Comment (zolomohan) [inline articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md:718]: Brief intro about the file.

File: articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md [hunk 15]
```diff
@@ -0,0 +1,832 @@
...
+}
+```
+
+### Running the App
+
+Now that we have successfully created a Real-time chat E2E App, the final step will be to run the server and the react app to test it.
+
+We need to note that our server runs on port 8000, and our frontend runs on port 3000. We need to [proxy](https://medium.com/bb-tutorials-and-thoughts/react-how-to-proxy-to-backend-server-5588a9e0347) the connection for our Node.js server to communicate with our frontend. To achieve this, we need to edit the React App package.json file located at `/chatfrontend/package.json` and add the below line of code:
+
+```
```
Comment (zolomohan) [inline articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md:801]:
```suggestion
```json
```

File: articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md [hunk 16]
```diff
@@ -0,0 +1,832 @@
...
+
+To run the frontend, type the below commands in the terminal:
+
+```bash
+$ cd chatfrontend
+$ npm start
+```
+
+The command will compile the whole project and run the React app; once this is complete, open the web browser and go to http://localhost:3000, type the user name and room name, also launch another tab and access http://localhost:3000 and, type a different user name but type same room name and test the app.
```
Comment (zolomohan) [inline articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md:823]:
```suggestion
The command will compile the project and run the React app. Once this is complete, open the web browser and go to http://localhost:3000, type a user name and a room name. 

Launch another tab and access http://localhost:3000 and, type a different user name but type same room name and test the app.
```

File: articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md [hunk 17]
```diff
@@ -0,0 +1,832 @@
...
+
+The video for testing our chat App is as shown below:
+
+<iframe width="853" height="460" src="https://www.youtube.com/embed/Nma-IAB2oSY" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
+
+### Conclusion
+In general, the app demonstrated above is quite simple and does not have many features a modern chat app would have. However, the idea, the code behind the app, and end-to-end encryption can be used to implement a real chat app. A lot more can be added from this chat app, but the concept and method can remain the same.
+
```
Comment (zolomohan) [inline articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md:831]:
```suggestion
In general, the app demonstrated above is quite simple and does not have many features a modern chat app would have. However, the idea, the code behind the app, and end-to-end encryption can be used to implement a real chat app.

```

**Review by zolomohan — CHANGES_REQUESTED** [review]

**Comment by zolomohan** [comment]
@ephnjor2021 Please go through the suggestions and ping me when you are done.

**Comment by ephnjor2021** [comment]
@zolomohan The requested changes have been made. 
Thank you

File: articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md [hunk 18]
```diff
@@ -0,0 +1,849 @@
...
+```typescript
+import { ProcessReducer } from "./process";
+import { combineReducers } from "redux";
+const rootReducers = combineReducers({
+  ProcessReducer: ProcessReducer,
+});
+
+export default rootReducers;
+```
```
Comment (zolomohan) [inline articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md:278]: Please provide a brief intro about these files and provide links to the concept of action and reducers.

**Review by zolomohan — CHANGES_REQUESTED** [review]
I noticed that most of the code snippets are long but none of them have comments to explain the code. Please add comments in all the code snippets to explain each statement.

**Comment by zolomohan** [comment]
@ephnjor2021 Please go through the suggestions and ping me when you are done.

**Comment by ephnjor2021** [comment]
@zolomohan The requested changes have been made.
Thank you.

**Comment by hectorkambow** [comment]
👋 Good afternoon - upon running your article through our 3rd party plagiarism checker it seemed to raise a few flags and the % was higher than we typically accept.
Please see attached PDF - and revisit the article to ensure we are contributing wholly unique and original content.
Be sure to see our resources page to see more info on plagiarism and what is considered as such.


@ephnjor2021 

[Creating a Simple Real-Time Chat App using React and Socket.io with E2E Encryption #2020.pdf](https://github.com/section-engineering-education/engineering-education/files/6354620/Creating.a.Simple.Real-Time.Chat.App.using.React.and.Socket.io.with.E2E.Encryption.2020.pdf)

**Comment by zolomohan** [comment]
@ephnjor2021 Please go through the attached document and try to reduce the overlap. Ping me when you are done.

**Comment by ephnjor2021** [comment]
@zolomohan The article overlap has been done and plagiarism reduced as possible.
Thank you.

**Review by zolomohan — CHANGES_REQUESTED** [review]
Hi @ephnjor2021. Upon conducting a Tech Review, I noticed the frontend of the application threw a minor error in `chat/chat.scss`.

Screenshot 2021-04-26 at 1 37 47 AM

When I commented out the scrollbars mixin, the app worked perfectly. 
Please fix this error.

**Comment by ephnjor2021** [comment]
@zolomohan The minor issue with css has been fixed.
Thank you

**Review by zolomohan — APPROVED** [review]

**Comment by hectorkambow** [comment]
OVerall good work @ephnjor2021 

please review my final edits as there were some issue and some parts that were difficult to read along. 

Please review to help avoid in future articles - also be sure to check our resources page for more tips.

cc @zolomohan

**Comment by ephnjor2021** [comment]
@hectorkambow Well noted. Thank you.

**Pull request merged**